cluster-api/apis/cluster/types.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package cluster contains types to represent Kubernetes cluster configuration.
+package cluster // import "k8s.io/kube-deploy/cluster-api/apis/cluster"
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// Cluster is an API object representing a cluster's control-plane
+// configuration and status.
+type Cluster struct {
+	metav1.ObjectMeta
+	metav1.TypeMeta
+
+	Spec   ClusterSpec
+	Status ClusterStatus
+}
+
+type ClusterSpec struct {
+	// Nominal version of Kubernetes control-plane to run.
+	KubernetesVersion KubernetesVersionInfo
+
+	// Basic API configuration
+	APIConfig APIServerConfig
+
+	// Cluster network configuration
+	ClusterNetwork ClusterNetworkingConfig
+
+	// Provider-specific serialized configuration to use during
+	// cluster creation. It is recommended that providers maintain
+	// their own versioned API types that should be
+	// serialized/deserialized from this field.
+	ProviderConfig string
+}
+
+type APIServerConfig struct {
+	// The address for the API server to advertise.
+	AdvertiseAddress string
+
+	// The port on which the API server binds.
+	Port uint32
+
+	// Extra Subject Alternative Names for the API server's serving cert.
+	ExtraSANs []string
+}
+
+type ClusterNetworkingConfig struct {
+	// The subnet from which service VIPs are allocated.
+	ServiceSubnet string
+
+	// The subnet from which POD networks are allocated.
+	PodSubnet string
+
+	// Domain name for services.
+	DNSDomain string
+}
+
+type KubernetesVersionInfo struct {
+	// Semantic version of Kubernetes to run.
+	Version string
+}
+
+type ClusterStatus struct {
+	// APIEndpoint represents the endpoint to communicate with the IP.
+	APIEndpoint APIEndpoint
+
+	// A simple boolean to indicate whether the control plane was
+	// successfully created.
+	Ready bool
```

Comment:
Shouldn't we avoid simple booleans and instead have a list of the different conditions?

Reply:
☝️ enumerated status's make sense to me

Reply:
Also I had a question here https://github.com/kubernetes/kube-deploy/pull/298#discussion_r147553493 about the expectation of the bool and what the criteria is behind it.

Reply:
It's noted in the Machines API proposal that the use of conditions in new APIs is discouraged: https://github.com/kubernetes/kube-deploy/pull/298/files#diff-a9cdbb84394e67bc6b18a54f950b838fR151 and will be avoided in the machines API at least.

It's probably best to be consistent in how we structure status between the two resources, regardless of how it's done 😄